Add functionality for new events

Leo-Send opened the pull request:
Adds functionality for the following events: 
- issue_type_added
- issue_type_changed
- issue_type_removed
- connected
- reopened
- parent_issue_added
- parent_issue_removed
- sub_issue_added
- sub_issue_removed

Leo-Send committed:
Add functionality to fetch external commits

external commits that reference an issue can now be fetched. They will
be marked with 'commitReferencesIssueExternal'. To that end, this commit
adds an 'external' field to each commit and a new method to extract
commits using the full url instead of just the relative one within the
repo.

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

Leo-Send committed:
Change 'extractHashtags' to ignore codeblocks

references to issues are now only found if they are ouside a code
environment (starting and ending with three '`'). This mirrors GitHubs
behaviour.

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

Leo-Send committed:
Document findings in 'README.md'

Add paragraph that documents intended and unintended behaviour for
'referenced' events

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

shirazJafri committed:
Add state reason and type to issues data

introduce new state reason and type data classes

Signed-off-by: Shiraz Jafri <shirazjafri@outlook.com>

shirazJafri committed:
Add event tracking for state changes

Add new event data subclass for state changed events

Signed-off-by: Shiraz Jafri <shirazJafri@outlook.com>

shirazJafri committed:
Add events tracking for issue type changed

Add new enum and eventdata subclasses to track changes in issue types

Signed-off-by: Shiraz Jafri <shirazJafri@outlook.com>

shirazJafri committed:
Change change_type field to not appear in result

no longer serializing change_type field in IssueTypeChangedEventData

Signed-off-by: Shiraz Jafri <shirazJafri@outlook.com>

shirazJafri committed:
Add events for parent and sub issues

Introduce new subclasses for parent and subissue events

Signed-off-by: Shiraz jarfi <shirazJafri@outlook.com>

shirazJafri committed:
Add tracking for connected events

Introduce new event data subclass for connected events

Signed-off-by: Shiraz Jafri <shirazJafri@outlook.com>

Leo-Send committed:
Add Section to README.md

reformat README and nested if statement

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

Leo-Send committed:
Fix spelling in 'README.md'

change BE words for consistency

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

Leo-Send committed:
Update Copyright headers

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

Leo-Send commented:
I have a few open questions/issues regarding this PR:
1. the 'connected' and 'subissue'/'parentissue' events do not contain any data on which issue the current issue is connected/related to. We could calculate this using timestamps, since both event types generate a counterpart in the connected/related issue. If we want to do this, I can look further into this.

2. Until now, I was unable to generate the 'sub_issue_changed' and 'parent_issue_changed' events. It seems to me that the only allowed interaction with subissues is adding or removing them. Should I still keep those events?

3. I am not sure about what I should do for the copyright headers, considering I only authored a change in one file so far.


Edit:
upon further investigation, all 'sub_issue' or 'parent_issue' event types do not appear in the REST api documentation. They do, however, appear in the graphQL api documentation and seem to behave the same in the REST api. But the 'sub_issue_changed' and 'parent_issue_changed' events do not exist there either.

Leo-Send committed:
Remove unnecessary map entry

key was duplicated with different values, overwriting previous entry

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

Leo-Send commented:
Quick update regarding connected events:
It is trivial to save the event id in the output json. Since this event id is the same for the connected events in both issues, it is then possible to find the related issue. The question in where we want to do that - in our toolchain, I think it would be sensible to do it in codeface-extraction or coronet. Adding this functionality to the GitHubWrapper would require a lot of work and I also do not think it fits here. It should be by far easiest in coronet, assuming we save the event id during the codeface-extraction issue processing, for example in event-info 1 or 2. If we do not want this id in the issues.list file, it can also be done in codeface-extraction.

bockthom commented:
Good idea!

I agree, it does not fit into the GitHubWrapper. So, just store it in the json file and we will find the related issue later.
While both codeface-extraction and coronet would be suitable for doing that, I'd prefer to do that in codeface-extraction, for two reasons:

(1) We do all the other processing in codeface-extraction. Why should we leave this one post-processing step to coronet? Would be inconsistent, I guess.
(2) When we would do this in coronet, we would need to recompute the connected issues every time we read issue data. Why should we do this every time and spend additional compute resources on that? If we do that in codeface-extraction, we do it only once, and the issues.list files contain already the correct linkage.

So, I'd suggest to compute the connected issue in codeface-extraction.

Leo-Send committed:
Improve SHA-1 extraction from issue comments

also add event ids and subissue numbers to data

SHA-1 extraction now only recognises hexadecimal strings length 7 and
above (previously 5)
Issue and Event data are extended by simple fields to hold event ids and
subissue numbers

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

Leo-Send committed:
Update copyright headers

also fix copyright headers in new files

Signed-off-by: Leo Sendelbach

Leo-Send commented:
This PR is ready for review, but should be merged after PR #31 is merged.

bockthom commented on src/de/uni_passau/fim/gitwrapper/EventDataProcessor.java:
```diff
@@ -43,7 +45,18 @@ class EventDataProcessor implements JsonDeserializer<EventData>, JsonSerializer<
+        map.put("closed", EventData.StateChangedEventData.class);
+        map.put("connected", EventData.ConnectedEventData.class);
+        map.put("reopened", EventData.StateChangedEventData.class);
+        map.put("issue_type_added", EventData.IssueTypeChangedEventData.class);
+        map.put("issue_type_changed", EventData.IssueTypeChangedEventData.class);
+        map.put("issue_type_removed", EventData.IssueTypeChangedEventData.class);
+        map.put("parent_issue_added", EventData.ParentIssueChangedEventData.class);
+        map.put("parent_issue_removed", EventData.ParentIssueChangedEventData.class);
+        map.put("parent_issue_changed", EventData.ParentIssueChangedEventData.class);
+        map.put("sub_issue_added", EventData.SubIssueChangedEventData.class);
+        map.put("sub_issue_removed", EventData.SubIssueChangedEventData.class);
+        map.put("sub_issue_changed", EventData.SubIssueChangedEventData.class);
```
If I remember correctly, a couple of weeks ago, we had a discussion in which you @Leo-Send mentioned that some/one of these "changed" events does not exist (not sure whether it was issue type, parent issue, or sub issue). Could you please check that again and remove the non-existing events here and everywhere else if applicable?

bockthom commented on src/de/uni_passau/fim/gitwrapper/EventDataProcessor.java:
```diff
@@ -43,7 +45,18 @@ class EventDataProcessor implements JsonDeserializer<EventData>, JsonSerializer<
         map.put("unlabeled", EventData.LabeledEventData.class);
         map.put("referenced", EventData.ReferencedEventData.class);
         map.put("merged", EventData.ReferencedEventData.class);
-        map.put("closed", EventData.ReferencedEventData.class);
+        map.put("closed", EventData.StateChangedEventData.class);
+        map.put("connected", EventData.ConnectedEventData.class);
+        map.put("reopened", EventData.StateChangedEventData.class);
```
Question 1: Which effect does it have that "closed" now uses the `StateChangedEventData` instead of `ReferencedEvent` data? Just want to make sure that we don't lose information when getting the new resolution type now.

Question 2: Why has "reopened" been added?  I assume because of the resolution type stuff. But how does "reopened" affect the resolution?

Request: Could you please sort the events in a way that closed and reopened are next to each other (and don't have "connected" in between).


Could you please figure that out?

bockthom commented on src/de/uni_passau/fim/gitwrapper/IssueTypeChangeEvent.java:
```diff
@@ -0,0 +1,67 @@
+     */
+    ADDED,
+
+    /**
+     * An issue type was changed for the issue.
+     */
+    CHANGED,
```
Same as above, please check whether `CHANGED` actually can occur.

bockthom commented on src/de/uni_passau/fim/gitwrapper/StateReason.java:
```diff
@@ -0,0 +1,80 @@
+     */
+    DUPLICATE,
+
+    /**
+     * Denotes that the state reason is not known or not specified.
+     * This can be used when the state reason is not applicable or not provided.
+     */
+    NONE,
+
+    /**
+     * Any state reason is included (can be used for unknown reasons).
+     */
+    ANY;
```
I remember that we had a discussion about what the difference between `NONE` and `ANY` is. I don't remember the outcome. Could you please check. Maybe we need to remove one of them...

And how do these enums transfer to the outcome in the resulting json files? Let's discuss that in our next meeting.

Leo-Send committed:
Remove superfluous state reasons and events

Parernt- and Subissues can not have a 'changed' event
None and Any state reasons were not achievable
Also removed the state reason field from issue data
StateChangedEvents can now have a commit (issues can be closed by a
commit)

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

bockthom commented on src/de/uni_passau/fim/gitwrapper/GitHubCommit.java:
```diff
@@ -119,4 +121,12 @@ public boolean isAddedToPullRequest() {
     }
+
+    void setExternal(boolean external) {
+        this.isExternal = true;
```
Needs to be addressed in PR #31

bockthom commented on src/de/uni_passau/fim/gitwrapper/IssueDataProcessor.java:
```diff
@@ -235,7 +265,17 @@ private List<String> extractSHA1s(String text) {
+                text = text + texts[i];
+            }
+        }
+
```
@hechtlC Do we need such a very lengthy explanation? I guess the commit hash lengths are kind of self-explanatory. I'll let this up to you whether we need documentation for that or not.

bockthom commented on src/de/uni_passau/fim/gitwrapper/IssueDataProcessor.java:
```diff
@@ -224,6 +232,28 @@ List<ReferencedLink<Integer>> parseIssues(IssueData issue, Gson gson) {
+        Optional<String> subIssueJson = repo.getJSONStringFromURL(issueBaseUrl + issue.number + "/sub_issues");
+
+        if (subIssueJson.isEmpty()) {
+            return new ArrayList<Integer>();
```
Any ideas on this?

bockthom commented:
I am satisfied with the changes in this PR. However, we also need to satisfy Copilot in this and the previous PR.
If the previous PR is updated, I can merge that one right away, such that this one can be rebased on top of the other one.

bockthom commented:
And please add you copyright header to StateReason.java, since you have modified it yourself now.

Leo-Send committed:
Rename field and setter for external commits

Getter previously did not use method parameter. Also renamed and added
docs to match existing methods.

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

Leo-Send committed:
Add exception if state reason is not a legal value

instead of assigning completed as default case, now throw an error. If
new state reasons are added later on githubs side, the occurence of this
error should prevent us from not noticing that change.

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

Leo-Send committed:
Improve performance  for extracting hashtags

using stringbuilder instead of appending, which would result in copying
the string

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

Leo-Send committed:
Add functionality to detect suggestions

suggestions contain a codeblock using the suggestion keyword after
opening the block. Add a field to reviewCommentData and result json for
this boolean

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

bockthom commented on src/de/uni_passau/fim/gitwrapper/ReferencedLinkProcessor.java:
````diff
@@ -64,6 +64,7 @@ public ReferencedLink deserialize(JsonElement json, Type typeOfT, JsonDeserializ
                 }
                 if (!json.getAsJsonObject().get("body").isJsonNull()) {
                     commentData.body = json.getAsJsonObject().get("body").getAsString();
+                    commentData.suggestion = commentData.body.contains("```suggestion") || commentData.body.contains("```suggestion");
````
Didn't we want to perform a case-independent check for "suggestion"?

bockthom commented on src/de/uni_passau/fim/gitwrapper/ReferencedLinkProcessor.java:
````diff
@@ -99,6 +100,7 @@ public ReferencedLink deserialize(JsonElement json, Type typeOfT, JsonDeserializ
                     commentData.commit_id = json.getAsJsonObject().get("commit_id").getAsString();
                     commentData.original_commit_id = json.getAsJsonObject().get("original_commit_id").getAsString();
                     commentData.body = json.getAsJsonObject().get("body").getAsString();
+                    commentData.suggestion = commentData.body.contains("```suggestion") || commentData.body.contains("```suggestion");
````
Same here

Leo-Send committed:
Resolve issues pointed out in review

Fixes:
- case sensitive check for suggestion
- minor spelling issue
- artifact from state reasons in issue data
- debug comment
- copyright headers

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

Leo-Send committed:
Add additional information to dummy user

If no userdata is found except for the username, initialize the dummy
with that username

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

bockthom commented on src/de/uni_passau/fim/gitwrapper/UserDataProcessor.java:
Please don't forget to update the copyright header here!

Leo-Send committed:
Update copyright header

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

Leo-Send committed:
Add support for locked events

This allows us to extract the reason for locking a conversation

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

Leo-Send committed:
Integrate requested changes

- update documentation, fix spelling
- add new fields to 'freeze' method
- use stringbuilder instead of concatenation
- fix deserialization of new event type
- change state change event to inherit from referenced event

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

bockthom commented on src/de/uni_passau/fim/gitwrapper/GitHubRepository.java:
```diff
@@ -1028,6 +1035,50 @@ Optional<GitHubCommit> getGithubCommit(String hash) {
+                 * JSON String, as it is not needed for further processing.
+                 */
+                LOG.info("Malformed JSON String when querying data for commit " + url + ". Neglect files element.");
+                String jsonStringFromURL = getJSONStringFromURL(url).get();
```
While the other comments of @Copilot make sense to me, I am not sure about this one.
@Leo-Send What do you think? Is this something that makes sense? I have the feeling that this is nonsense. The printed log warning seems to be unrelated to the `isPresent` call that Copilot suggests. Or do you perceive it differently?

Leo-Send replied:
The case described by copilot can only happen if the api returns null (since every other return should be castable to a string), which I believe would be an indicator of the api being completely unavailable. I do not think we should continue the execution in this case.

bockthom commented on src/de/uni_passau/fim/gitwrapper/UserDataProcessor.java:
```diff
@@ -76,9 +77,20 @@ public UserData deserialize(JsonElement json, Type typeOfT, JsonDeserializationC
-            return DUMMY_USER;
+            if (username == null || username.isEmpty()) {
+                LOG.warning("Could not get information about unknown user!");
+                (guess ? guessedUsersByUsername : strictUsersByUsername).put(username, DUMMY_USER);
```
I agree with copilot here. We should not put `null` or `""` into the dicct.

Leo-Send committed:
Integrate requested changes

Remove unnecessary type check, refactor into new helperfunction, and
formatting

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

bockthom reviewed (approved):
Everything seems to work well in my test run, so let's assume everything is fine.
Thanks for the numerous enhancements to GitHubWrapper @Leo-Send.
